GetStream/stream-chat-android
# [2951] Handle mentions in DiffUtil

https://github.com/GetStream/stream-chat-android/issues/2951

### 🎯 Goal

Handle mentions in DiffUtil so that the changes are propagated to the UI.

### 🛠 Implementation details

Added handling of the following Message fields in DiffUtil: `mentionedUsersIds`, `mentionedUsers`

### 🎨 UI Changes

The screen recordings below show the result of applying the patch from the "testing" section below:

<table>
<thead>
<tr>
<th>Before</th>
<th>After</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<video src="https://user-images.githubusercontent.com/9600921/150318990-3f9c7e2d-0ecf-40d5-ac96-7277b0bf7abe.mp4" controls="controls" muted="muted" />
</td>
<td>
<video src="https://user-images.githubusercontent.com/9600921/150318961-d566b6ac-6744-4fe3-b4b1-0f68ea74073d.mp4" controls="controls" muted="muted" />
</td>
</tr>
</tbody>
</table>

### 🧪 Testing

1. Apply the patch below
2. Send a message with a mention
3. Observe that the `Message.mentionedUsers` field is not null

<details>
  <summary>Patch file to apply these changes for testing</summary>

```
Index: stream-chat-android-ui-components-sample/src/main/kotlin/io/getstream/chat/ui/sample/application/App.kt
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/stream-chat-android-ui-components-sample/src/main/kotlin/io/getstream/chat/ui/sample/application/App.kt b/stream-chat-android-ui-components-sample/src/main/kotlin/io/getstream/chat/ui/sample/application/App.kt
--- a/stream-chat-android-ui-components-sample/src/main/kotlin/io/getstream/chat/ui/sample/application/App.kt	(revision 2cb10db793d0bf20c244f73dbadb31e70693f4c6)
+++ b/stream-chat-android-ui-components-sample/src/main/kotlin/io/getstream/chat/ui/sample/application/App.kt	(date 1642666781098)
@@ -2,12 +2,16 @@
 
 import android.app.Application
 import android.os.Build
+import android.widget.TextView
 import coil.Coil
 import coil.ImageLoader
 import coil.decode.GifDecoder
 import coil.decode.ImageDecoderDecoder
+import com.getstream.sdk.chat.adapter.MessageListItem
 import io.getstream.chat.android.client.utils.internal.toggle.ToggleService
 import io.getstream.chat.android.core.internal.InternalStreamChatApi
+import io.getstream.chat.android.ui.ChatUI
+import io.getstream.chat.android.ui.transformer.ChatMessageTextTransformer
 import io.getstream.chat.ui.sample.BuildConfig
 import io.getstream.chat.ui.sample.data.user.SampleUser
 import io.getstream.chat.ui.sample.data.user.UserRepository
@@ -34,6 +38,15 @@
         )
         ApplicationConfigurator.configureApp(this)
         initializeToggleService()
+
+        ChatUI.messageTextTransformer =
+            ChatMessageTextTransformer { textView: TextView, messageItem: MessageListItem.MessageItem ->
+                val messageText = messageItem.message.text
+                val upstreamMentionedUserId = messageItem.message.mentionedUsersIds.firstOrNull()
+                val downstreamMentionedUserId = messageItem.message.mentionedUsers.firstOrNull()?.id
+
+                textView.text = "$messageText [$upstreamMentionedUserId $downstreamMentionedUserId]"
+            }
     }
 
     private fun getApiKey(): String {

```

</details>

### ☑️Contributor Checklist

#### General

#### Code & documentation
~~- [ ] New code is covered by unit tests~~
~~- [ ] Affected documentation updated (KDocs, docusaurus, tutorial)~~

### ☑️Reviewer Checklist
